libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/lifecycle/ViewLifecycleRegistry.kt

```diff
@@ -45,22 +46,45 @@ class ViewLifecycleRegistry(
     }
 
     private val attachStateChangeListener = object : View.OnAttachStateChangeListener {
-        override fun onViewAttachedToWindow(p0: View?) {
-            currentState = hostingLifecycleOwner.lifecycle.currentState
-            isAttached = true
+        override fun onViewAttachedToWindow(p0: View) {
+            doOnAttached(p0)
         }
 
-        override fun onViewDetachedFromWindow(p0: View?) {
-            isAttached = false
-            if (hostingLifecycleOwner.lifecycle.currentState.isAtLeast(State.STARTED)) {
-                currentState = State.CREATED
-            }
+        override fun onViewDetachedFromWindow(p0: View) {
+            doOnDetached()
         }
     }
 
     init {
-        hostingLifecycleOwner.lifecycle.addObserver(hostingLifecycleObserver)
         view.addOnAttachStateChangeListener(attachStateChangeListener)
+        if (view.isAttachedToWindow) {
+            doOnAttached(view)
+        }
+    }
+
+    private fun doOnAttached(view: View) {
+        if (isAttached) return
+
+        hostingLifecycleOwner?.lifecycle?.removeObserver(hostingLifecycleObserver)
+
+        val hostingLifecycleOwner = ViewTreeLifecycleOwner.get(view)
+            ?: throw IllegalStateException(
+                "Please ensure that the hosting activity/fragment is a valid LifecycleOwner"
+            )
+        currentState = hostingLifecycleOwner.lifecycle.currentState
+        hostingLifecycleOwner.lifecycle.addObserver(hostingLifecycleObserver)
+        this@ViewLifecycleRegistry.hostingLifecycleOwner = hostingLifecycleOwner
+        isAttached = true
+    }
+
+    private fun doOnDetached() {
+        if (!isAttached) return
+
+        isAttached = false
+        val hostingLifecycleOwner: LifecycleOwner = checkNotNull(hostingLifecycleOwner)
+        if (hostingLifecycleOwner.lifecycle.currentState.isAtLeast(State.STARTED)) {
+            currentState = State.CREATED
+        }
```

Comment:
Wonder if it's worth to unsubscribe lifecycle and attach listeners at some point (in onDestroy?)

Reply:
Is there a scenario in which we could leak here? I don't clearly see how the registry could outlive the view or the hosting lifecycle owner to make us leak.

Reply:
My thought here was that registry and view could have shorter lifetime that hosting fragment/activity (thus their lifecycle). E.g. if view was removed dynamically from the layout, while fragment/activity continue to live.

Reply:
> Wonder if it's worth to unsubscribe lifecycle and attach listeners at some point (in onDestroy?)

You do not need to unregister lifecycle observers onDestroy because the lifecycle framework handles that

Reply:
> Changing to isAtLeast(State.CREATED)

This whole check in general makes no difference, a local lifecycle cannot be further than the hosting lifecycle and `CREATED` is the lowest we track anyway. For example, we can't be in a situation where the fragment is `CREATED` but the view (local) is `RESUMED`. Just calling `currentState = State.CREATED` without any checks works in all situation that have correct usage.

That said, I'm going to leave it as a safeguard in case something goes terribly wrong and the view is attached to the window before the parent lifecycle is alive. I don't think that can truly happen but the check makes sense from the perspective of the `ViewLifecycleRegistry` which doesn't know the greater context (for example the fact that the view cannot be attached before lifecycle is created).

> My thought here was that registry and view could have shorter lifetime that hosting fragment/activity (thus their lifecycle). E.g. if view was removed dynamically from the layout, while fragment/activity continue to live.

That doesn't seem to be the problem because we don't keep the reference to the `view` in the `ViewLifecycleRegistry`, and the Android `LifecycleRegistry` only holds a weak reference to the owning lifecycle. So `hostingLifecycleObserver` that we have subscribed (and could outlive the view) doesn't hold onto anything meaningful, and we need to keep it subscribed anyway to deliver the final `onDestroy` after which it's cleaned up as @kmadsen mentioned.

Reply:
> That doesn't seem to be the problem because we don't keep the reference to the view in the ViewLifecycleRegistry

Oh, indeed, somehow I've thought it was `val view` hence keeping ref to the view.